### DISCUSSION TITLE
RoutedEventAttribute: How to handle a custom event args type?

### DISCUSSION THREAD

**Discussion by fubar-coder** [HavenDV/DependencyPropertyGenerator#61] [Q&A]
Hi,

the dependency property source generator stuff really works well. After that, I wanted to try the routed events, but I stumbled over the following problem:

Let's assume that I have an event handler delegate, like `EventHandler<MyCustomEventArgs>`, with `MyCustomEventArgs` being defined as:

```c#
public class MyCustomEventArgs : RoutedEventArgs
{
  public MyCustomEventArgs(int myValue)
  {
    MyValue = myValue;
  }

  public MyCustomEventArgs(int myValue, RoutedEvent evt)
    : base(evt)
  {
    MyValue = myValue;
  }

  public MyCustomEventArgs(int myValue, RoutedEvent evt, object? source)
    : base(evt, source)
  {
    MyValue = myValue;
  }

  public int MyValue { get; }
}
```

Now, I'd like to use the following attribute:

```c#
[RoutedEvent<EventHandler<MyCustomEventArgs>>("MyCustomEvent", RoutedEventStrategy.Bubble)]
```

The problem is, that it automatically generates a `OnMyCustomEvent`, which creates a `RoutedEventArgs` instead of the needed `MyCustomEventArgs`.

Is there a way to provide the correct event arguments?

**Comment by HavenDV** [comment]
Hi. As far as I remember, this is not supported, I did not expect such use during development. But I think it is easy to add

**Reply by fubar-coder** [reply]
Maybe the `RoutedEventAttribute` is the wrong tool. What I'd really like to have is an event that gets triggered before/after a dependency property will be/was changed, maybe even using `RoutedPropertyChangedEventArgs<T>`?

EDIT: Support for `RoutedPropertyChangedEventArgs<T>` might be (very) nice as a first step?

**Reply by HavenDV** [reply]
You need to use partial methods like this:
```cs
[DependencyProperty<bool>(""IsSpinning2"")]
public partial class MyControl : UserControl
{
    // Optional
    partial void OnIsSpinning2Changing(bool newValue)
    {
    }
    
    // Optional
    partial void OnIsSpinning2Changed(bool newValue)
    {
    }
}
```

**Reply by fubar-coder** [reply]
I already use those callbacks, but I was (also) looking for an easy way to reduce the boilerplate code for routed events and a way to pass the new value to this event.

TBH, maybe the easiest "fix" would be a way to disable the automatically generated `OnMyCustomEvent` function, so that I can provide my own, while keeping all other generated code?
